### ISSUE TITLE
Index searchable data

### ISSUE THREAD

**Issue by jsdir** [martinshkreli/shchem#10]
- Determine the fields that we will be searching by. Use https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/cid/1650002/record/JSON for reference.

### Indexed properties

| field                  | `urn.label` | `urn.name`     |
|------------------------|-------------|----------------|
| `iupac_name_allowed`     | IUPAC Name  | Allowed        |
| `iupac_name_cas`         | IUPAC Name  | CAS-like Style |
| `iupac_name_preferred`   | IUPAC Name  | Preferred      |
| `iupac_name_systematic`  | IUPAC Name  | Systematic     |
| `iupac_name_traditional` | IUPAC Name  | Traditional    |
| `inchi_standard`         | InChI       | Standard       |
| `inchi_key_standard`     | InChIKey    | Standard       |
| `smiles_isomeric`        | SMILES      | Isomeric       |
| `smiles_canonical`       | SMILES      | Canonical      |

**Comment by jsdir** [comment]
Even though we can index on a JSON field in postgres, there's a complication that makes it a little more difficult.

Inside the compound record listed above, the props are not keyed and are instead contained in an array.

```json
{
  "props": [
    {
      "urn": {
        "label": "IUPAC Name",
        "name": "Allowed",
        "datatype": 1,
        "version": "2.2.0",
        "software": "LexiChem",
        "source": "openeye.com",
        "release": "2016.09.28"
      },
      "value": {
        "sval": "[(1S)-1-methyl-2-(2-methylanilino)-2-oxo-ethyl] (E)-3-(3-nitrophenyl)prop-2-enoate"
      }
    },
    {
      "urn": {
        "label": "IUPAC Name",
        "name": "CAS-like Style",
        "datatype": 1,
        "version": "2.2.0",
        "software": "LexiChem",
        "source": "openeye.com",
        "release": "2016.09.28"
      },
      "value": {
        "sval": "(E)-3-(3-nitrophenyl)-2-propenoic acid [(2S)-1-(2-methylanilino)-1-oxopropan-2-yl] ester"
      }
    }
  ]
}
```

We have several options for this:

1. Create a function in postgres that finds the prop value given its label or name, and have postgres [index on an expression](https://www.postgresql.org/docs/9.4/static/indexes-expressional.html) that uses that function. eg. `CREATE INDEX compound_iupac ON compounds (get_compound_prop(compound, "IUPAC Name", "Allowed"));`
2. Change the seed script to transform the JSON data by extracting the prop values and adding them to a top-level hash that can be easily indexed
3. Add specific columns to the compound table and change the seed script to pull the prop values into those columns

@gabelerner Any thoughts?

**Comment by gabelerner** [comment]
1) is the fastest/easiest.  might need to do some raw sql on the node side for the where clause but whatever.  making it work first.
2) this is too close to doing 1 but takes an extra step ... we still will have a list of indices that reference deep into json ... so my least favorite choice
3) i would do this via a view of the original table + selects on the json into new columns. views can have indices and sequelize should be able to treat them just like tables.  postgres is really, really good at this and it shouldn't require reseeding all the time.

basically avoiding changing seed script since we'd constantly have to reseed/etc if we had that process.  hard to choose but 1 is fastest for now and then we can go to 3 later if it's not working.

**Comment by jsdir** [comment]
Using a view sounds like the better choice. It will also simplify the work on the client since we will have defined, top-level attributes rather than just of a list of props. We already have the list of fields to be collected so I'm going to assign you to this issue.
